src/main/scala/nl/vroste/zio/kinesis/client/DynamicConsumer.scala

```diff
@@ -110,13 +112,104 @@ object DynamicConsumer {
         )
     )
 
+  final case class RecordProcessor[R, T](value: Record[T] => RIO[R, Unit]) extends (Record[T] => RIO[R, Unit]) {
+    def apply(r: Record[T]): RIO[R, Unit] = value(r)
+  }
+
+  /**
+   * Similar to `shardedStream` accessor but provides the `recordProcessor` callback function for processing records
+   * and takes care of checkpointing. The other difference is that it returns a ZIO of unit rather than a ZStream.
+   *
+   * @param streamName Name of the Kinesis stream
+   * @param applicationName Application name for coordinating shard leases
+   * @param deserializer Deserializer for record values
+   * @param requestShutdown Effect that when completed will trigger a graceful shutdown of the KCL
+   *   and the streams.
+   * @param initialPosition Position in stream to start at when there is no previous checkpoint
+   *   for this application
+   * @param isEnhancedFanOut Flag for setting retrieval config - defaults to `true`. If `false` polling config is set.
+   * @param leaseTableName Optionally set the lease table name - defaults to None. If not specified the `applicationName` will be used.
+   * @param workerIdentifier Identifier used for the worker in this application group. Used in logging
+   *   and written to the lease table.
+   * @param maxShardBufferSize The maximum number of records per shard to store in a queue before blocking
+   *   the KCL record processor until records have been dequeued. Note that the stream returned from this
+   *   method will have internal chunk buffers as well.
+   * @param checkpointBatchSize Maximum number of records before checkpointing
+   * @param checkpointDuration Maximum interval before checkpointing
+   * @param recordProcessor A function for processing a `Record[T]`
+   * @tparam R ZIO environment type required by the `deserializer` and the `recordProcessor`
+   * @tparam T Type of record values
+   * @return ZIO of unit
```

Comment:
Might be good to indicate that this ZIO only completes when record processing is stopped via `requestShutdown` or fails when the consumer stream fails.

Reply:
:+1: